src/poetry/puzzle/provider.py

```diff
@@ -878,7 +916,20 @@ def _resolve_overlapping_markers(
         dependencies = self._merge_dependencies_by_constraint(dependencies)
 
         new_dependencies = []
+
+        # We have can sort out the implicit "not required" dependency determined
+        # in _add_implicit_dependencies, because it does not overlap with
+        # any other dependency for sure.
+        for i, dep in enumerate(dependencies):
+            if dep.constraint.is_empty():
+                new_dependencies.append(dependencies.pop(i))
```

Comment:
The `list.pop` method can be very slow operation, and I think that it can be avoided here, by using a "blacklist" approach, e.g.

```python
blacklist = set()
for dep in dependencies:
    if dep.constraint.is_empty():
        blacklist.add(dep)
        break
```

Then later on in `itertools.product` use 
`repeat=len(dependencies) - len(blacklist)`.
And when looping over `dep in dependencies` again, simply skip it if `dep in blacklist`.

This avoids the `list.pop` operation, which has a time-complexity `O(n)`, by relying on `set.__contains__`, which is only `O(1)`.